File: frontend/types/jsa.ts
import { z } from "zod";
import { APDSchema } from "./work-permit"; // reuse APD definitions

export const HazardSourceSchema = z.enum([
  "Mesin",
  "Peralatan",
  "Lingkungan Kerja",
  "Material",
  "OPL/Metode Kerja",
  "Perilaku",
]);

export const JSAAnalysisStepSchema = z.object({
  work_step: z.string().min(1, "Urutan pekerjaan wajib diisi"),
  hazard_source: z.string().min(1, "Sumber bahaya wajib diisi"),
  potential_hazard: z.string().min(1, "Potensi bahaya wajib diisi"),
  control_measure: z.string().min(1, "Upaya pengendalian wajib diisi"),
});

export const JSAFormSchema = z.object({
  job_number: z.string().min(1, "No Job wajib diisi").optional(), // set by BE
  job_name: z.string().min(1, "Nama Pekerjaan wajib diisi"),
  date: z.string(), // ISO format
  supervisor: z.string().min(1, "Nama Pengawas wajib diisi"),
  department: z.string().min(1, "Departemen wajib diisi"),
  executor: z.string().min(1, "Nama Pelaksana wajib diisi"),
  required_apd: z.array(APDSchema),
  hazard_sources: z.array(HazardSourceSchema).min(1, "Pilih minimal 1 sumber bahaya"),
  analysis_steps: z.array(JSAAnalysisStepSchema).min(1, "Minimal 1 langkah analisis"),
});

export type JSAForm = z.infer<typeof JSAFormSchema>;

export interface JSAAnalysisStep extends z.infer<typeof JSAAnalysisStepSchema> {
  id?: number;
  sequence: number;
}

export interface JSA extends JSAForm {
  id: number;
  job_number: string;
  analysis_steps: JSAAnalysisStep[];
  status: "DRAFT" | "WAITING_MANAGER" | "WAITING_K3" | "APPROVED_K3" | "REJECTED";
  created_at: string;
}

File: backend/internal/domain/jsa.go
package domain

import (
	"time"

	"gorm.io/datatypes"
)

// JSAStatus enumeration representing scalable workflow
const (
	JSAStatusDraft          = "DRAFT"
	JSAStatusWaitingManager = "WAITING_MANAGER" // Future-proof scalable step
	JSAStatusWaitingK3      = "WAITING_K3"
	JSAStatusApproved       = "APPROVED_K3"
	JSAStatusRejected       = "REJECTED"
)

// JSA represents Form F.120.01.00.11 Rev.01
type JSA struct {
	ID           uint      `gorm:"primaryKey" json:"id"`
	JobNumber    string    `gorm:"uniqueIndex;not null" json:"job_number"`
	JobName      string    `gorm:"not null" json:"job_name"`
	Date         time.Time `gorm:"not null" json:"date"`
	Supervisor   string    `json:"supervisor"`
	Department   string    `json:"department"`
	Executor     string    `json:"executor"`
	WorkPermitID *uint     `json:"work_permit_id"` // Optional link to Ijin Kerja

	// APD (Helm, Kacamata, dll)
	RequiredAPD datatypes.JSON `json:"required_apd"`

	// Sumber Bahaya Umum (Mesin, Peralatan, dll)
	HazardSources datatypes.JSON `json:"hazard_sources"`

	// Analysis Steps (Tabel Analisis Dinamis)
	AnalysisSteps []JSAAnalysisStep `gorm:"foreignKey:JSAID" json:"analysis_steps"`

	// Workflow
	Status        string     `gorm:"default:'WAITING_K3';not null" json:"status"`
	
	// Scalable Approvers
	ManagerID     *uint      `json:"manager_id"`
	ManagerApprov *time.Time `json:"manager_approved_at"`

	K3OfficerID     *uint      `json:"k3_officer_id"`
	K3OfficerApprov *time.Time `json:"k3_officer_approved_at"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// JSAAnalysisStep represents a single dynamic row in the safety analysis process
type JSAAnalysisStep struct {
	ID             uint   `gorm:"primaryKey" json:"id"`
	JSAID          uint   `gorm:"not null;index" json:"jsa_id"`
	Sequence       int    `gorm:"not null" json:"sequence"` // Order 1, 2, 3...
	WorkStep       string `gorm:"not null" json:"work_step"` // Urutan pekerjaan
	HazardSource   string `json:"hazard_source"`             // Sumber bahaya spesifik langkah
	PotentialHazard string `json:"potential_hazard"`         // Potensi Bahaya
	ControlMeasure string `json:"control_measure"`           // Upaya Pengendalian
}
